app/src/main/java/com/zegoggles/smssync/mail/MessageGenerator.java

```diff
@@ -144,7 +162,28 @@ class MessageGenerator {
             sentDate = new Date();
         }
         final int msg_box = toInt(msgMap.get("msg_box"));
-        headerGenerator.setHeaders(msg, msgMap, DataType.MMS, details.address, details.getRecipient(), sentDate, msg_box);
+
+        // We could thread by contact ID, not by thread ID. Original author thought this value was more stable.
+        // It works pretty badly with MMS threads though.
+//        String mmsThreadId = details.getRecipient();
+
+        // We could grab all the raw addresses (phone numbers), and sort/unique/join them for gmail's thread id
+        // This doesn't work well, because sometimes phone numbers have a "+1" prefix, sometimes not.
+//        Set<String> rawAddresses = new HashSet<>();
+//        rawAddresses.addAll(details.rawAddresses);
+//        List<String> rawAddressesSorted = new ArrayList<>(rawAddresses.size());
+//        rawAddressesSorted.addAll(rawAddresses);
+//        Collections.sort(rawAddressesSorted);
+//        String mmsThreadId = stringJoin(rawAddressesSorted, "-");
+
+        // We could thread by the messaging app's thread ID. The original author thought this wasn't very stable,
+        // but it's the best option for MMS so far, and it looks good from limited testing.
+        String mmsThreadId = msgMap.get(Telephony.BaseMmsColumns.THREAD_ID);
+
+        // Tip - if you want to try different strategies for computing mmsThreadId against the same messages,
+        // make sure to generate a new MESSAGE_ID as well. It seems like gmail caches REFERENCES if you reuse the same MESSAGE_ID.
```

Comment:
Some context, from what I learned doing this: MESSAGE_ID is the unique message on gmail's end. REFERENCES is the field (on gmail's end) that determines which email an email is replying to.

For the last message grouping fix I built, I wasn't seeing it take effect, even after I deleted all my sms backups from gmail and ran the back again. I was able to determine that gmail was caching the REFERENCES field by clicking "show original" in gmail and looking at the message header. I had to force the app to generate new MESSAGE_IDs. Basically there's some string we pass into the hash function that generates MESSAGE_ID. You can just concatenate "v2", "v3", etc. to the hash input. It's almost like bumping a version number. Once I generated a new MESSAGE_ID for those messages I had already backed up before, **then** my last fix started working. 

However, this has repercussions for releasing this fix. I tried backing up all the messages on my phone with a clean build of this current PR. I'm using a different gmail label (`SMS-testing` instead of the usual `SMS`) to keep things clean, but the app was still generating the same MESSAGE_ID as before. So when I checked on some older (already backed up) messages, they weren't grouped correctly 😢 

I can vet this out, but I believe if we version/change the hash input (like we start concatenating "v2"), that will force gmail to recognize these as new messages. So then users will be able to reset the app state and re-backup their existing messages, and then older threads should get fixed 🎉 I expect the old messages would linger around in gmail as well, so there may some duplicates... but users will only get this if they choose to re-backup their existing messages.

This will also be a useful opportunity for me to check against more test messages. I'll check this out some day soon!

Reply:
Good news! Versioning the hash input https://github.com/jberkel/sms-backup-plus/pull/1097/commits/ab80cc0788714231fe1f25d615ebe9c375ff0224 fixes this problem nicely. This will be helpful to force gmail to recognize these messages as new threads and group them correctly.